app/src/main/java/com/raulh82vlc/functionalkotlinarrow/data/network/NetworkDataSource.kt

```diff
@@ -0,0 +1,134 @@
+
+object NetworkDataSource {
+
+    private const val JSON_FLICKR_FEED_LITERAL = "jsonFlickrFeed("
+    private const val JSON_FLICKR_FEED_KEYWORD = "jsonFlickrFeed\\("
+    private const val API_HOST = "https://api.flickr.com/services/"
+
+    val gson = Gson()
+
+    val retrofit = Retrofit.Builder()
+            .baseUrl(API_HOST)
+            .client(createOkHttp())
+            .addCallAdapterFactory(CoroutineCallAdapterFactory())
+            .build()
+            .create(FeedApi::class.java)
+
+    private fun createOkHttp(): OkHttpClient {
+        val okHttpClient = OkHttpClient.Builder()
+                .connectTimeout(3, TimeUnit.SECONDS)
+                .writeTimeout(10, TimeUnit.SECONDS)
+                .readTimeout(10, TimeUnit.SECONDS)
+        return okHttpClient.build()
+    }
+
+    fun requestFeed(format: String): IO<List<FeedItemCacheModel>> {
+        val monadError = IO.monadError()
+        return monadError.binding {
+            val result = runInAsyncContext(
+                    f = { transformToRepository(retrofit.getFeed(format)) },
+                    onError = { monadError.raiseError<List<FeedItemCacheModel>>(it) },
+                    onSuccess = { monadError.just(it) },
+                    AC = IO.async()
+            ).bind()
+            result.bind()
+        }.fix()
+    }
+
+    private fun transformToRepository(feed: Call<ResponseBody>): List<FeedItemCacheModel> {
+        val response = execute(feed)?.string()
+        response?.run {
+            if (hasFeedFormat(this)) {
+                return NetworkToCacheMapper.map(deserializeFeedJSON(extractJSONFromResponse(this)).feedItems)
+            }
+            return emptyList()
+        }
+        return emptyList()
+    }
+
+    private fun hasFeedFormat(responseBodyResult: String): Boolean =
+        responseBodyResult.contains(JSON_FLICKR_FEED_LITERAL)
+
+    private fun extractJSONFromResponse(response: String): String {
+        val responseOutput = response.replaceFirst(JSON_FLICKR_FEED_KEYWORD.toRegex(), "")
+        return responseOutput.substring(0, responseOutput.length - 1)
+    }
+
+    private fun deserializeFeedJSON(responseJson: String): FeedApiModel {
+        return gson.fromJson(responseJson, FeedApiModel::class.java)
+    }
+
+    /**
+     * Run in Async context a Coroutine
+     * Thanks to Jorge Castillo for this method
+     * package com.github.jorgecastillo.kotlinandroid.io.algebras.persistence
+     */
+    private fun <F, A, B> runInAsyncContext(
+            f: () -> A,
+            onError: (Throwable) -> B,
+            onSuccess: (A) -> B, AC: Async<F>): Kind<F, B> {
+        return AC.async { proc ->
+            kotlinx.coroutines.experimental.async(CommonPool) {
+                val result = Try { f() }.fold(onError, onSuccess)
+                proc(result.right())
+            }
+        }
+    }
+
+    @Throws(ApiException::class)
+    fun execute(call: Call<ResponseBody>): ResponseBody? {
```

Comment:
Instead of returning a `nullable` in here you can return `Try<ResponseBody>` and it doesn't need the `@Throws` annotation.
```
fun Call<ResponseBody>.tryToExecute(): Try<ResponseBody> = 
  Try { execute() }
    .recoverWith(Try.raise(ApiException("Network error", e)))
    .flatMap { response ->
      response?.takeIf { it.isSuccessful }?.body() ?: throw ApiException(response.errorBody().toString(), null)
    }
```

Reply:
I have to learn a lot about this using Arrow, cheers for this

Reply:
will do for a new refactor, not for this one